src/coreclr/jit/flowgraph.cpp

```diff
@@ -4252,6 +4252,80 @@ FlowGraphDfsTree* Compiler::fgComputeDfs()
 template FlowGraphDfsTree* Compiler::fgComputeDfs<false>();
 template FlowGraphDfsTree* Compiler::fgComputeDfs<true>();
 
+//------------------------------------------------------------------------
+// fgComputeLoopAwareDfs: Compute a depth-first search tree for the flow graph
+// where in the RPO traversal, loop bodies are visited before loop successors.
+//
+// Returns:
+//   The tree.
+//
+// Notes:
+//   If the flow graph has loops, the DFS will be reordered such that loop bodies are compact.
+//   This will invalidate BasicBlock::bbPreorderNum and BasicBlock::bbPostorderNum.
+//
+FlowGraphDfsTree* Compiler::fgComputeLoopAwareDfs()
```

Comment:
Is there anything gained from trying to represent this as an actual `FlowGraphDfsTree`? I think it would make more sense to have a utility function that given `FlowGraphDfsTree` and `FlowGraphNaturalLoops` visits the blocks in RPO that respects the loop structure. It would basically be a slight generalization of what we have in VN already.

The "compute DFS tree" into "identify loops" into "now create another DFS tree" seems wasteful an conceptually a bit odd.

Reply:
> Is there anything gained from trying to represent this as an actual FlowGraphDfsTree?

Probably not.

> I think it would make more sense to have a utility function that given FlowGraphDfsTree and FlowGraphNaturalLoops visits the blocks in RPO that respects the loop structure.

That sounds sensible -- I'll try modeling this after `FlowGraphNaturalLoop::VisitLoopBlocksReversePostOrder`.